akka-stream/src/main/scala/akka/stream/scaladsl/KillSwitch.scala

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright (C) 2015-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import akka.Done
+import akka.stream._
+import akka.stream.scaladsl.KillSwitch.KillableGraphStageLogic
+import akka.stream.stage._
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * Creates shared or single kill switches which can be used to control completion of graphs from the outside.
+ *  - The factory ``shared()`` returns a [[SharedKillSwitch]] which provides a [[Flow]] that can be
+ *    used in arbitrary number of graphs and materializations. The switch simultaneously
+ *    controls completion in all of those graphs.
+ *  - The factory ``single()`` returns a [[Flow]] that materializes to a [[UniqueKillSwitch]] which always unique
+ *    to that materialized Flow itself.
+ *
+ * Creates a [[SharedKillSwitch]] that can be used to externally control the completion of various streams.
+ *
+ */
+object KillSwitch {
+
+  /**
+   * Creates a new [[SharedKillSwitch]] with the given name that can be used to control the completion of multiple
+   * streams from the outside simultaneously.
+   *
+   * @see SharedKillSwitch
+   */
+  def shared(name: String): SharedKillSwitch = new SharedKillSwitch(name)
+
+  /**
+   * Creates a new [[Flow]] that materializes to an external switch that allows external completion of that unique
+   * materialization. Different materializations result in different, independent switches.
+   *
+   * For a Bidi version see [[KillSwitch#singleBidi]]
+   */
+  def single[T]: Flow[T, T, UniqueKillSwitch] = uniqueKillSwitchFlow.asInstanceOf[Flow[T, T, UniqueKillSwitch]]
+
+  /**
+   * Creates a new [[BidiFlow]] that materializes to an external switch that allows external completion of that unique
+   * materialization. Different materializations result in different, independent switches.
+   *
+   * For a Flow version see [[KillSwitch#single]]
+   */
+  def singleBidi[T1, T2]: BidiFlow[T1, T1, T2, T2, UniqueKillSwitch] =
+    uniqueKillSwitchBidi.asInstanceOf[BidiFlow[T1, T1, T2, T2, UniqueKillSwitch]]
+
+  private val uniqueKillSwitchFlow: Flow[Any, Any, UniqueKillSwitch] = Flow.fromGraph(UniqueKillSwitch)
+  private val uniqueKillSwitchBidi: BidiFlow[Any, Any, Any, Any, UniqueKillSwitch] = BidiFlow.fromGraph(BidiBreaker)
+
+  abstract class KillableGraphStageLogic(val terminationSignal: Future[Done], _shape: Shape) extends GraphStageLogic(_shape) {
+    override def preStart(): Unit = {
+      val cb = getAsyncCallback[Try[Done]](onSwitch)
```

Comment:
why create this here if you are going to call completeStage/failStage on the next few lines?